### ISSUE TITLE
Debug build does not set `-DPy_DEBUG` with cmake

### ISSUE THREAD

**Issue by SimonHeybrock** [pybind/pybind11#2018]
### TLDR
I think the cmake variable `PYTHON_IS_DEBUG` is not visible inside the scope of the function `pybind11_add_module`.

### What

Trying to create a debug build, using the debug version of Python with `cmake`:

- `-DCMAKE_BUILD_TYPE=Debug`
- `-DPYTHON_EXECUTABLE=/usr/lib/python3.6-dgb`

Build goes fine, but importing the resulting module gives an `Import Error` about the undefined symbol `PyModule_Create2`.

From what I have read, this indicates that we are linking correctly against the debug version of Python, but the includes are wrong, since `PyModule_Create2` has a different name in the debug version of Python.

### Solution?

- https://github.com/pybind/pybind11/blob/37352491225358b97ce302273bf2d887a477efb0/tools/pybind11Tools.cmake#L142-L147 is not working as it should.
- I have verified that `PYTHON_IS_DEBUG` is set correctly in https://github.com/pybind/pybind11/blob/master/tools/FindPythonLibsNew.cmake#L115, and it is also correct in `pybind11Tools.cmake` (outside the function scope)...
- ... but **not when the function is called**.

I am having trouble understanding `cmake`'s scoping rules. If I set `PYTHON_IS_DEBUG` in the cmake CACHE it works. I cannot tell whether this is the correct solution though?

### Versions

- cmake 3.10.2
- pybind11 2.4.3
- python3.6-dbg
- Ubuntu
